elements/lisk-tree/src/merkle_tree/merkle_tree.ts

```diff
@@ -267,6 +270,141 @@ export class MerkleTree {
+			indexes[binaryIndex] = index;
+		}
+
+		const siblingHashes = [];
+		let currentNode: NodeInfo | undefined;
+		let currentNodeHash = Buffer.alloc(0);
+
+		while(!(Object.keys(indexes)[0] === undefined)) {
+			const currentLocation = indexes[Object.keys(indexes)[0]];
+			const {
+				nodeIndex: currentNodeIndex,
+				layerIndex: currentLayerIndex,
+			} = currentLocation;
+			const binaryIndex = getBinaryString(currentNodeIndex, this._getHeight() - currentLayerIndex).toString();
+
+			try {
+				currentNodeHash = (await this._locationToHashMap.get(
+					getBinaryString(currentNodeIndex, this._getHeight() - currentLayerIndex),
+				)) as Buffer;
+				currentNode = await this.getNode(currentNodeHash);
+			} catch (err) {
+				delete indexes[binaryIndex];
+				continue;
+			}
+
+			const {
+				layerIndex: pairLayerIndex,
+				nodeIndex: pairNodeIndex,
+			} = getPairLocation({
+				layerIndex: currentNode.layerIndex,
+				nodeIndex: currentNode.nodeIndex,
+				size: this._size,
+			});
+			const pairNodeLocation: NodeLocation = {
+				layerIndex: pairLayerIndex,
+				nodeIndex: pairNodeIndex,
+			};
+
+			const pairBinaryIndex = getBinaryString(pairNodeIndex, this._getHeight() - pairLayerIndex).toString();
+
+			const pairNodeHash = (await this._locationToHashMap.get(
+				getBinaryString(pairNodeIndex, this._getHeight() - pairLayerIndex),
+			)) as Buffer;
+
+			if (indexes[pairBinaryIndex]) {
+				delete indexes[pairBinaryIndex];
+			} else {
+				siblingHashes.push(pairNodeHash);
+			}
+
+			delete indexes[binaryIndex];
+
+			const parentIndex: NodeLocation = getParentLocation(currentLocation, pairNodeLocation);
+			const {
+				layerIndex: parentLayerIndex,
+				nodeIndex: parentNodeIndex,
+			} = parentIndex;
+			const parentBinaryIndex = getBinaryString(parentNodeIndex, this._getHeight() - parentLayerIndex).toString();
+
+			if (parentBinaryIndex !== '0') {
+				indexes[parentBinaryIndex] = parentIndex;
+			}
+		}
+
+		return siblingHashes;
+	}
+
+	public async update(idxs: ReadonlyArray<number>, updateData: ReadonlyArray<Buffer>) {
+		const updateHashes = [];
+
+		for (const data of updateData) {
+			const leafValueWithoutNodeIndex = Buffer.concat(
+				[LEAF_PREFIX, data],
+				LEAF_PREFIX.length + data.length,
+			);
+			const leafHash = hash(leafValueWithoutNodeIndex);
+			updateHashes.push(leafHash);
+		}
+
+		const pairHashes = await this.getSiblingHashes(idxs);
+		const calculatedTree = calculatePathNodes(updateHashes, this._size, idxs, pairHashes);
+		
+		for (const updatedNode of Object.values(calculatedTree)) {
+				const existingNodeHash = await this._locationToHashMap.get(
+					getBinaryString(updatedNode.nodeIndex, this._getHeight() - updatedNode.layerIndex),
+				);
+				await this._locationToHashMap.set(
+					getBinaryString(updatedNode.nodeIndex, this._getHeight() - updatedNode.layerIndex),
+					updatedNode.hash,
+				);
+				await this._hashToValueMap.set(updatedNode.hash, updatedNode.value);
```

Comment:
This should get the node value from updateData I guess

Reply:
done

Reply:
when we update `hashToValue` we should also update `locationToHash`